pkg/daemon/ceph/client/pool.go

```diff
@@ -418,6 +421,78 @@ func CreateReplicatedPoolForApp(context *clusterd.Context, clusterInfo *ClusterI
 	}
 
 	logger.Infof("creating replicated pool %s succeeded", poolName)
+
+	if checkFailureDomain {
+		if err = ensureFailureDomain(context, clusterInfo, clusterSpec, poolName, pool); err != nil {
+			return nil
+		}
+	}
+	return nil
+}
+
+func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, clusterSpec *cephv1.ClusterSpec, poolName string, pool cephv1.PoolSpec) error {
+	if pool.FailureDomain == "" {
+		logger.Debugf("skipping check for failure domain on pool %q as it is not specified", poolName)
+		return nil
+	}
+
+	logger.Debugf("checking that pool %q has the failure domain %q", poolName, pool.FailureDomain)
+	details, err := GetPoolDetails(context, clusterInfo, poolName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get pool %q details", poolName)
+	}
+
+	// Find the failure domain for the current crush rule
+	rule, err := getCrushRule(context, clusterInfo, details.CrushRule)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get crush rule %q", details.CrushRule)
+	}
+	currentFailureDomain := extractFailureDomain(rule)
+	if currentFailureDomain == pool.FailureDomain {
+		logger.Debugf("pool %q has the expected failure domain %q", poolName, pool.FailureDomain)
+		return nil
+	}
+	if currentFailureDomain == "" {
+		logger.Warningf("failure domain not found for crush rule %q, proceeding to create a new crush rule", details.CrushRule)
+	}
+
+	// Use a crush rule name that is unique to the desired failure domain
+	crushRuleName := fmt.Sprintf("%s_%s", poolName, pool.FailureDomain)
+	logger.Infof("updating pool %q failure domain from %q to %q with new crush rule %q", poolName, currentFailureDomain, pool.FailureDomain, crushRuleName)
+	logger.Infof("crush rule %q will no longer be used by pool %q", details.CrushRule, poolName)
+
+	// Create a new crush rule for the expected failure domain
+	if err := createReplicationCrushRule(context, clusterInfo, clusterSpec, crushRuleName, pool); err != nil {
+		return errors.Wrapf(err, "failed to create replicated crush rule %q", crushRuleName)
+	}
+
+	// Update the crush rule on the pool
+	if err := setCrushRule(context, clusterInfo, poolName, crushRuleName); err != nil {
+		return errors.Wrapf(err, "failed to set crush rule on pool %q", poolName)
+	}
+
+	logger.Infof("Successfully updated pool %q failure domain to %q", poolName, pool.FailureDomain)
+	return nil
+}
+
+func extractFailureDomain(rule ruleSpec) string {
```

Comment:
@leseb Any suggestions on detecting the crush rule failure domain other than this? This code path only applies to a basic replica pool, and not to EC, stretch, or pools with sub-failure domains. At least this should be more robust with the for loop than the previous implementation expecting the failure domain to be at `rule.Steps[1]`